packages/slate-locales-plugin/index.js

```diff
@@ -0,0 +1,183 @@
+
+module.exports = class localesPlugin {
+  constructor(options) {
+    this.options = options;
+  }
+  apply(compiler) {
+    compiler.hooks.compilation.tap('Slate Locales Plugin', this.addLocales);
+  }
+
+  async addLocales(compilation) {
+    const dirents = await fs.readdir(
+      path.resolve(compilation.options.context, 'sections'),
+      {withFileTypes: true},
+    );
+    // eslint-disable-next-line prefer-arrow-callback
+    dirents.forEach((dirent) => {
+      if (dirent.isDirectory()) {
+        traverseDirectory(
+          dirent,
+          path.resolve(compilation.options.context, 'sections'),
+          compilation,
+          dirents,
+        );
+      } else if (dirent.isFile() && path.extname(dirent.name) === '.liquid') {
+        getSchema(
+          dirent,
+          path.resolve(compilation.options.context, 'sections'),
+          compilation,
+          dirents,
+        );
+      }
+    });
+  }
+};
+
+async function getSchema(liquidDirent, directoryPath, compilation, dirents) {
+  if (liquidDirent.name === 'template.liquid') {
+    const outputKey = `../sections${
+      directoryPath.split('/sections')[1]
+    }.liquid`;
+
+    if (
+      dirents.find((obj) => obj.name === 'locales' && obj.isDirectory()) &&
+      dirents.find((obj) => obj.name === 'schema.json' && obj.isFile())
+    ) {
+      compilation.assets[outputKey] = new ConcatSource(
+        new RawSource(
+          fs.readFileSync(path.resolve(directoryPath, 'template.liquid')),
+        ),
+        new RawSource(
+          `{% schema %}\n ${JSON.stringify(
+            createMainSchema(
+              combineLocales(path.resolve(directoryPath, 'locales')),
+              path.resolve(directoryPath, 'schema.json'),
+            ),
+            null,
+            2,
+          )}{% endschema %}`,
+        ),
+      );
+    } else if (
+      dirents.find((obj) => obj.name === 'schema.json' && obj.isFile())
+    ) {
+      compilation.assets[outputKey] = new ConcatSource(
+        new RawSource(
+          fs.readFileSync(path.resolve(directoryPath, 'template.liquid')),
+        ),
+        new RawSource(
+          `{% schema %}\n${fs.readFileSync(
+            path.resolve(directoryPath, 'schema.json'),
+          )}{% endschema %}`,
+        ),
+      );
+    } else {
+      compilation.assets[outputKey] = new RawSource(
+        fs.readFileSync(path.resolve(directoryPath, 'template.liquid')),
+      );
+    }
+  } else {
+    const liquidContent = await fs.readFile(
+      path.resolve(directoryPath, liquidDirent.name),
+      'utf8',
+    );
+    const outputKey = `../sections${
+      directoryPath.split('/sections')[1] === ''
+        ? `/${liquidDirent.name}`
+        : `${directoryPath.split('/sections')[1]}/${liquidDirent.name}`
+    }`;
+    if (
+      liquidContent.match(/{%\s*schema\s*%}([\s\S]+){%\s*endschema\s*%}/gim) ||
+      !dirents.find(
+        (obj) => obj.name === liquidDirent.name.replace('.liquid', '.json'),
```

Comment:
What's the reason for looking for a json file of the same name?

Reply:
The functionality https://github.com/Shopify/slate/pull/954 this PR wanted